src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,245 @@
+            if (cachedLimiters is null)
+            {
+                // Timer already running, do nothing
+                // This might be an indication that there is a slow ReplenishingRateLimiter.TryReplenish implementation
+                // Or many limiters that need updating and a short timer period
+                // Or threadpool exhaustion unrelated to this type directly
+                // We might want to write an event for this in the future so it's visible to diagnostic tools
+                return;
+            }
+
+            lock (limiter.Lock)
+            {
+                if (limiter._disposed)
+                {
+                    cachedLimiters.Clear();
+                    return;
+                }
+
+                // If the cache has been invalidated we need to recreate it
+                if (limiter._cacheInvalid)
+                {
+                    bool cacheStillInvalid = false;
+                    foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                    {
+                        if (kvp.Value.IsValueCreated)
+                        {
+                            if (kvp.Value.Value is ReplenishingRateLimiter)
+                            {
+                                cachedLimiters.Add(kvp.Value);
+                            }
+                        }
+                        else
+                        {
+                            // In rare cases the RateLimiter will be added to the storage but not be initialized yet
+                            // keep cache invalid if there was a non-initialized RateLimiter
+                            // the next time we run the timer the cache will be updated
+                            // with the initialized RateLimiter
+                            cacheStillInvalid = true;
+                        }
+                    }
+                    limiter._cacheInvalid = cacheStillInvalid;
+                }
+            }
+
+            try
+            {
+                // cachedLimiters is safe to use outside the lock because it is only updated by the Timer
+                // and the Timer avoids re-entrancy issues via the _executingTimer field
+                foreach (Lazy<RateLimiter> rateLimiter in cachedLimiters)
+                {
+                    Debug.Assert(rateLimiter.IsValueCreated && rateLimiter.Value is ReplenishingRateLimiter);
+                    ((ReplenishingRateLimiter)rateLimiter.Value).TryReplenish();
+                }
+            }
+            finally
+            {
+                Interlocked.Exchange(ref limiter._cachedLimiters, cachedLimiters);
+                TaskCompletionSource<bool>? tcs = Interlocked.Exchange(ref limiter._timerTask, null);
+                // DisposeAsync was called and is waiting for the Timer to indicate completion
+                if (tcs is not null)
+                {
+                    tcs.SetResult(true);
```

Comment:
I'm pretty sure `Timer.DisposeAsync()` already does something similar to this to wait on any ongoing timer callbacks to complete. Can we just rely on that?

https://github.com/dotnet/runtime/blob/8f7761af3763d0b84acefbb117ed6becd83061d1/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs#L635-L643

Reply:
No, it's netstandard2.1+